### ISSUE TITLE
no-unreachable false positive with hoisted functions that end with a semicolon

### ISSUE THREAD

**Issue by weswigham** [palantir/tslint#527]
``` ts
function f9() {
    hoisted();
    return 0;

    function hoisted() {
        return 0;
    };
}
```

Triggers a `no-unreachable` on the end of the hoisted function.

Technically there is _actually_ unreachable code - there's an empty 0 character statement between the function declaration and the semicolon, since only function expressions would normally have ASI. (I think.) I wouldn't normally consider forbidding semicolons after function declarations as part of the `no-unreachable` rule, though, so this is probably considered a bug.

**Comment by adidahiya** [comment]
according to https://github.com/Microsoft/TypeScript/pull/5048, a related issue occurs with local interface declarations and type aliases. not sure if we need separate issues for those

**Comment by IllusionMH** [comment]
Probably can be closed, since checks for unreachable code are now rely on `no-unreachable` compiler flag.
